source/Ship.cpp

```diff
@@ -3374,6 +3375,66 @@ double Ship::IdleHeat() const
 	double dissipation = HeatDissipation() + activeCooling / MaximumHeat();
 	if(!dissipation) return production ? numeric_limits<double>::max() : 0;
 	return production / dissipation;
+	*/
+
+	// The Secant Method, closely related to Newton's Method,
+	// is a robust approximation, approaching the zero of any curve
+	// quickly and cheaply. It does not provide infinite accuracy,
+	// but in exchange it is trivial to add more functionality to
+	// the curve it is approximating a zero of.
+	// To do so- implement your heat change in ship::DoGeneration()
+	// and then add the same change to Ship::NetIdleHeatAt().
+
+	int attempts = 10;
+	double firstGuess = 0.;
+	double firstOutput = NetIdleHeatAt(firstGuess);
+
+	double secondGuess = MaximumHeat();
+	double secondOutput = NetIdleHeatAt(secondGuess);
+
+	double middlingGuess;
+	double middlingOutput;
+
+	// Guard clause, to catch weird cases.
+	if(firstGuess == secondGuess || firstOutput <= secondOutput)
+		return secondOutput > 0. ? numeric_limits<double>::max() : 0;
+
+	// 0.001 here is a magic number which could be anything "sufficiently small".
+	// Normally it's called an Epsilon, a number that approaches but is not zero.
+	while((attempts > 0) && (fabs(firstGuess - secondGuess) > .001))
```

Comment:
```suggestion
	for(int attempts = 0; (attempts < 10) && (fabs(firstGuess - secondGuess) > .001); ++atempts)
```
Just a suggestion, nothing that you have to do.

Reply:
Two comments on why this change is good:  1) it'll let you remove "attempts" from the higher scope, avoiding accidental re-use of the variable,  and 2) it avoids a possible infinite loop.

That being said though, you've written the code so the changes aren't necessary. Little else happens in the function, so (1) isn't a problem. Also, the only way your fabs(f-s)>.001 could go wrong is if the numbers are so large that a result <=.001 is impossible. 

Hence, the change isn't really necessary; it's just a matter of "best practices," but this is a game and not pacemaker code or a class assignment.